File: src/low_level/pattern.jl
export get_size, set_size, get_edge, set_edge, has_directed_path, has_cycle
export is_DAG, to_DAG, set, to_network, has_incoming_edge, has_outgoing_edge
export print, get_adjacent_nodes, get_parents, get_children
export DSL_EDGETYPE_NONE, DSL_EDGETYPE_UNDIRECTED, DSL_EDGETYPE_DIRECTED

# ------

const DSL_EDGETYPE_NONE       = unsafe_load(cglobal((:EDSL_EDGETYPE_NONE,       LIB_SMILE), Cint))
const DSL_EDGETYPE_UNDIRECTED = unsafe_load(cglobal((:EDSL_EDGETYPE_UNDIRECTED, LIB_SMILE), Cint))
const DSL_EDGETYPE_DIRECTED   = unsafe_load(cglobal((:EDSL_EDGETYPE_DIRECTED,   LIB_SMILE), Cint))

# ------

get_size( pat::Pattern ) = ccall( (:pattern_getSize, LIB_SMILE), Cint, (Ptr{Void},), pat.ptr )
set_size( pat::Pattern, size::Integer ) = ccall( (:pattern_setSize, LIB_SMILE), Void, (Ptr{Void},Cint), pat.ptr, size )
get_edge( pat::Pattern, from::Integer, to::Integer ) = ccall( (:pattern_getEdge, LIB_SMILE), Cint, (Ptr{Void},Cint,Cint), pat.ptr, from, to )
set_edge( pat::Pattern, from::Integer, to::Integer, edgetype::Cint ) = ccall( (:pattern_setEdge, LIB_SMILE), Void, (Ptr{Void},Cint,Cint,Cint), pat.ptr, from, to, edgetype )
has_directed_path( pat::Pattern, from::Integer, to::Integer ) = ccall( (:pattern_hasDirectedPath, LIB_SMILE), Bool, (Ptr{Void},Cint,Cint), pat.ptr, from, to )
has_cycle( pat::Pattern ) = ccall( (:pattern_hasCycle, LIB_SMILE), Bool, (Ptr{Void},), pat.ptr )
is_DAG( pat::Pattern ) = ccall( (:pattern_isDAG, LIB_SMILE), Bool, (Ptr{Void},), pat.ptr )
to_DAG( pat::Pattern ) = ccall( (:pattern_toDAG, LIB_SMILE), Bool, (Ptr{Void},), pat.ptr )
set( pat::Pattern, net::Network ) = ccall( (:pattern_set, LIB_SMILE), Void, (Ptr{Void},Ptr{Void}), pat.ptr, net.ptr )
to_network( pat::Pattern, dset::Dataset, net::Network ) = ccall( (:pattern_toNetwork, LIB_SMILE), Bool, (Ptr{Void},Ptr{Void},Ptr{Void}), pat.ptr, dset.ptr, net.ptr )
has_incoming_edge( pat::Pattern, to::Integer ) = ccall( (:pattern_hasIncomingEdge, LIB_SMILE), Bool, (Ptr{Void},Cint), pat.ptr, to )
has_outgoing_edge( pat::Pattern, from::Integer ) = ccall( (:pattern_hasOutgoingEdge, LIB_SMILE), Bool, (Ptr{Void},Cint), pat.ptr, from )
print( pat::Pattern ) = ccall( (:pattern_print, LIB_SMILE), Void, (Ptr{Void},), pat.ptr )

function get_adjacent_nodes( pat::Pattern, node::Integer )
	size = get_size(pat)
	arr  = Array(Cint, size)
	len  = ccall( (:pattern_getAdjacentNodes, LIB_SMILE), Cint, (Ptr{Void},Cint,Ptr{Cint}), pat.ptr, node, arr )
	arr[1:len]
end
function get_parents( pat::Pattern, node::Integer )
	size = get_size(pat)
	arr  = Array(Cint, size)
	len  = ccall( (:pattern_getParents, LIB_SMILE), Cint, (Ptr{Void},Cint,Ptr{Cint}), pat.ptr, node, arr )
	arr[1:len]
end
function get_children( pat::Pattern, node::Integer)
	size = get_size(pat)
	arr  = Array(Cint, size)
	len  = ccall( (:pattern_getChildren, LIB_SMILE), Cint, (Ptr{Void},Cint,Ptr{Cint}), pat.ptr, node, arr )
	arr[1:len]
end

File: src/wrapper/cpp/pattern.cpp
// SMILE.JL
// A C library wrapper for SMILE (Structural Modeling, Inference, and Learning Engine)
// so that users may call SMILE functions through Julia
//
// author: Tim Wheeler
// date: October 2014
// Stanford Intelligence Systems Laboratory

///////////////////////////////////////
//              PATTERN              //
///////////////////////////////////////

int EDSL_EDGETYPE_NONE       = DSL_pattern::None;
int EDSL_EDGETYPE_UNDIRECTED = DSL_pattern::Undirected;
int EDSL_EDGETYPE_DIRECTED   = DSL_pattern::Directed;

void * createPattern() {
	DSL_pattern* pat = new DSL_pattern();
	void * retval = pat;
	return retval;
}
void freePattern( void * void_pattern ) {
	DSL_pattern * pat = reinterpret_cast<DSL_pattern*>(void_pattern);
	delete pat;
}
int pattern_getSize( void * void_pattern ) {
	DSL_pattern * pat = reinterpret_cast<DSL_pattern*>(void_pattern);
	return pat->GetSize();
}
void pattern_setSize( void * void_pattern, int size ) {
	DSL_pattern * pat = reinterpret_cast<DSL_pattern*>(void_pattern);
	pat->SetSize(size);
}
int pattern_getEdge( void * void_pattern, int from, int to ) {
	DSL_pattern * pat = reinterpret_cast<DSL_pattern*>(void_pattern);
	return int(pat->GetEdge(from, to));
}
void pattern_setEdge( void * void_pattern, int from, int to, int edgetype ) {
	DSL_pattern * pat = reinterpret_cast<DSL_pattern*>(void_pattern);
	DSL_pattern::EdgeType type = DSL_pattern::None;
	if (edgetype == int(DSL_pattern::Undirected))
		type = DSL_pattern::Undirected;
	else if (edgetype == int(DSL_pattern::Directed))
		type = DSL_pattern::Directed;
	pat->SetEdge(from, to, type);
}
bool pattern_hasDirectedPath( void * void_pattern, int from, int to ) {
	DSL_pattern * pat = reinterpret_cast<DSL_pattern*>(void_pattern);
	return pat->HasDirectedPath( from, to );
}
bool pattern_hasCycle( void * void_pattern ) {
	DSL_pattern * pat = reinterpret_cast<DSL_pattern*>(void_pattern);
	return pat->HasCycle();
}
bool pattern_isDAG( void * void_pattern ) {
	DSL_pattern * pat = reinterpret_cast<DSL_pattern*>(void_pattern);
	return pat->IsDAG();
}
bool pattern_toDAG( void * void_pattern ) {
	DSL_pattern * pat = reinterpret_cast<DSL_pattern*>(void_pattern);
	return pat->ToDAG();
}
void pattern_set( void * void_pattern, void * void_net ) {
	DSL_pattern * pat = reinterpret_cast<DSL_pattern*>(void_pattern);
	DSL_network * net = reinterpret_cast<DSL_network*>(void_net);
	pat->Set(*net);
}
bool pattern_toNetwork( void * void_pattern, void * void_dset, void * void_net ) {
	DSL_pattern * pat  = reinterpret_cast<DSL_pattern*>(void_pattern);
	DSL_dataset * dset = reinterpret_cast<DSL_dataset*>(void_dset);
	DSL_network * net  = reinterpret_cast<DSL_network*>(void_net);
	return pat->ToNetwork(*dset, *net);
}
bool pattern_hasIncomingEdge( void * void_pattern, int to ) {
	DSL_pattern * pat = reinterpret_cast<DSL_pattern*>(void_pattern);
	return pat->HasIncomingEdge(to);
}
bool pattern_hasOutgoingEdge( void * void_pattern, int from ) {
	DSL_pattern * pat = reinterpret_cast<DSL_pattern*>(void_pattern);
	return pat->HasOutgoingEdge(from);
}
void pattern_print( void * void_pattern ) {
	DSL_pattern * pat = reinterpret_cast<DSL_pattern*>(void_pattern);
	pat->Print();
}
int pattern_getAdjacentNodes( void * void_pattern, int node, int * arr ) {
	DSL_pattern * pat = reinterpret_cast<DSL_pattern*>(void_pattern);
	std::vector<int> adj;
	pat->GetAdjacentNodes(node, adj);
	for (unsigned i = 0; i < adj.size(); i++)
		arr[i] = adj[i];
	return (int)(adj.size());
}
int pattern_getParents( void * void_pattern, int node, int * arr ) {
	DSL_pattern * pat = reinterpret_cast<DSL_pattern*>(void_pattern);
	std::vector<int> par;
	pat->GetParents(node, par);
	for (unsigned i = 0; i < par.size(); i++)
		arr[i] = par[i];
	return (int)(par.size());
}
int pattern_getChildren( void * void_pattern, int node, int * arr ) {
	DSL_pattern * pat = reinterpret_cast<DSL_pattern*>(void_pattern);
	std::vector<int> child;
	pat->GetChildren(node, child);
	for (unsigned i = 0; i < child.size(); i++)
		arr[i] = child[i];
	return (int)(child.size());
}